incubator/hnc/pkg/controllers/object_controller_new.go

```diff
@@ -38,25 +59,306 @@ type ObjectReconcilerNew struct {
 
 	// GVK is the group/version/kind handled by this reconciler.
 	GVK schema.GroupVersionKind
+
+	// Affected is a channel of event.GenericEvent (see "Watching Channels" in
+	// https://book-v1.book.kubebuilder.io/beyond_basics/controller_watches.html) that is used to
+	// enqueue additional objects that need updating.
+	Affected chan event.GenericEvent
 }
 
 // +kubebuilder:rbac:groups=*,resources=*,verbs=get;list;watch;create;update;patch;delete
 
 func (r *ObjectReconcilerNew) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	resp := ctrl.Result{}
+	ctx := context.Background()
 	log := r.Log.WithValues("trigger", req.NamespacedName)
-	log.Info("New object controller placeholder")
-	return resp, nil
+
+	log.Info("Reconciling object")
+
+	// Read the object. If it's missing, assume it's been deleted or not yet created.
+	inst := &unstructured.Unstructured{}
+	inst.SetGroupVersionKind(r.GVK)
+	inst.SetNamespace(req.Namespace)
+	inst.SetName(req.Name)
+	found := true
+	if err := r.Get(ctx, req.NamespacedName, inst); err != nil {
+		if !errors.IsNotFound(err) {
+			log.Info("Couldn't read")
+			return resp, err
+		}
+		found = false
+	}
+	act, so := r.syncWithForest(ctx, log, found, inst)
+
+	return resp, r.operate(ctx, log, act, inst, so)
+}
+
+// syncWithForest syncs the object instance with the in-memory forest. This is the only place we should
```

Comment:
Nit - maybe move everything about the lock to line 102 where we actually lock/unlock the mutex? It's a bit too detailed if I'm just trying to figure out what this function does. But the return values are critical. It might also be worth mentioning that this doesn't just _return_ data - it can also _update_ the forest (eg if a source object is added or removed).